File: app/src/main/java/de/bikebean/app/ui/drawer/history/position/PositionHistoryFragment.java
package de.bikebean.app.ui.drawer.history.position;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import java.util.ArrayList;
import java.util.List;

import de.bikebean.app.R;
import de.bikebean.app.db.state.LocationState;
import de.bikebean.app.db.state.State;
import de.bikebean.app.ui.drawer.history.HistoryAdapter;
import de.bikebean.app.ui.drawer.history.HistoryFragment;
import de.bikebean.app.ui.drawer.history.HistoryViewModel;
import de.bikebean.app.ui.drawer.sms_history.SmsViewModel;

public class PositionHistoryFragment extends HistoryFragment {

    @Override
    public @NonNull View onCreateView(final @NonNull LayoutInflater inflater,
                             final @Nullable ViewGroup container,
                             final @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        final @NonNull View v = inflater.inflate(R.layout.fragment_history_position, container, false);

        positionHistoryButton = v.findViewById(R.id.positionButton);
        batteryHistoryButton = v.findViewById(R.id.batteryButton);

        recyclerView = v.findViewById(R.id.recyclerView2);
        noDataText = v.findViewById(R.id.noDataText);

        positionHistoryButton.setOnClickListener(super::navigateToTab);
        batteryHistoryButton.setOnClickListener(super::navigateToTab);

        return v;
    }

    @Override
    protected @NonNull HistoryViewModel getNewStateViewModel() {
        return new ViewModelProvider(this).get(PositionHistoryViewModel.class);
    }

    @Override
    protected void setupListeners() {
        final @NonNull SmsViewModel smsViewModel = new ViewModelProvider(this).get(SmsViewModel.class);
        smsViewModel.getAllIds().observe(this, this::updateStates);
    }

    @Override
    protected @NonNull HistoryAdapter getNewAdapter(final @NonNull Context ctx) {
        return new PositionHistoryAdapter(ctx,
                ((PositionHistoryViewModel) st).getLocationStates().getValue()
        );
    }

    private void updateStates(final @NonNull List<Integer> smsIdList) {
        final @NonNull List<LocationState> locationStates = new ArrayList<>();

        new Thread(() -> {
            for (int smsId : smsIdList) {
                final @Nullable LocationState locationState =
                        updateLocationStates(((PositionHistoryViewModel) st).getAllLocation(smsId));
                if (locationState != null)
                    locationStates.add(locationState);
            }

            ((PositionHistoryViewModel) st).setLocationsState(locationStates);
        }).start();

        ((PositionHistoryViewModel) st).getLocationStates().removeObservers(this);
        ((PositionHistoryViewModel) st).getLocationStates().observe(this, this::setStatesToAdapter);
    }

    private @Nullable LocationState updateLocationStates(final @NonNull List<State> states) {
        State latState = null;
        State lngState = null;
        State accState = null;
        State noCellTowersState = null;
        State noWifiAccessPointsState = null;

        if (states.size() < 5)
            return null;

        for (State s : states)
            switch (State.KEY.getValue(s)) {
                case LAT:
                    latState = s;
                    break;
                case LNG:
                    lngState = s;
                    break;
                case ACC:
                    accState = s;
                    break;
                case NO_CELL_TOWERS:
                    noCellTowersState = s;
                    break;
                case NO_WIFI_ACCESS_POINTS:
                    noWifiAccessPointsState = s;
                    break;
            }

        if (latState != null && lngState != null && accState != null
                && noCellTowersState != null && noWifiAccessPointsState != null)
            return new LocationState(
                    latState, lngState, accState,
                    noCellTowersState, noWifiAccessPointsState
            );

        return null;
    }
}

File: app/src/main/java/de/bikebean/app/ui/drawer/status/location/history/PositionHistoryFragment.kt
package de.bikebean.app.ui.drawer.status.location.history

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import de.bikebean.app.R
import de.bikebean.app.db.state.LocationState
import de.bikebean.app.db.state.State
import de.bikebean.app.db.state.State.KEY
import de.bikebean.app.ui.drawer.sms_history.SmsViewModel
import de.bikebean.app.ui.drawer.status.history.HistoryAdapter
import de.bikebean.app.ui.drawer.status.history.HistoryFragment
import de.bikebean.app.ui.drawer.status.history.HistoryViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class PositionHistoryFragment : HistoryFragment() {

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View = inflater
            .inflate(R.layout.fragment_history_position, container, false)
            .apply {
                recyclerView = findViewById(R.id.recyclerView2)
                noDataText = findViewById(R.id.noDataText)
            }.also {
                super.onCreateView(inflater, container, savedInstanceState)
            }

    override val newStateViewModel: HistoryViewModel get() =
            ViewModelProvider(this).get(PositionHistoryViewModel::class.java)

    override fun setupListeners() {
        ViewModelProvider(this).get(SmsViewModel::class.java).allIds
                .observe(this, ::updateStates)
    }

    override fun getNewAdapter(ctx: Context): HistoryAdapter = PositionHistoryAdapter(ctx,
            (st as PositionHistoryViewModel).locationStates.value
    )

    private fun updateStates(smsIdList: List<Int>) {
        val st = st as PositionHistoryViewModel
        lifecycleScope.launch {
            withContext(Dispatchers.Default) {
                st.setLocationsState(
                        List(smsIdList.size) {
                            updateLocationStates(st.getAllLocation(smsIdList[it]))
                        }.filterNotNull()
                )
            }
        }
        st.locationStates.removeObservers(this)
        st.locationStates.observe(this, ::setStatesToAdapter)
    }

    private fun updateLocationStates(states: List<State>): LocationState? = listOfNotNull(
            states.firstOrNull(KEY::isLat),
            states.firstOrNull(KEY::isLng),
            states.firstOrNull(KEY::isAcc),
            states.firstOrNull(KEY::isNoCellTowers),
            states.firstOrNull(KEY::isNoWifiAccessPoints)
    ).let { list ->
        if (list.size == 5) LocationState(
                list.component1(),
                list.component2(),
                list.component3(),
                list.component4(),
                list.component5()
        ) else null
    }
}